auth-jwt-service/src/main/java/io/stargate/auth/jwt/AuthJwtService.java

```diff
@@ -69,6 +79,148 @@ public StoredCredentials validateToken(String token) throws UnauthorizedExceptio
+    JSONObject stargateClaims = extractClaimsFromJWT(token);
+
+    preCheckSchemaWrite(stargateClaims, keyspace, table);
+
+    // Just return the result. No value in doing a post check since we can't roll back anyway.
+    return action.call();
+  }
+
+  private JSONObject extractClaimsFromJWT(String token) throws JSONException {
+    // Grab the custom claims from the JWT. It's safe to work with the JWT as a plain Base64 encoded
+    // json object here since by this point we've already authenticated the request.
+    String[] parts = token.split("\\.");
+    String decodedPayload = new String(Base64.getUrlDecoder().decode(parts[1]));
+    JSONObject payload = new JSONObject(decodedPayload);
+    return payload.getJSONObject(CLAIMS_FIELD);
+  }
+
+  private void preCheckSchemaRead(JSONObject stargateClaims, String keyspace, String table) {
+    // TODO: [doug] 2020-11-2, Mon, 17:23 Finish implementing
+  }
+
+  private void preCheckSchemaWrite(JSONObject stargateClaims, String keyspace, String table) {
+    // TODO: [doug] 2020-11-2, Mon, 17:23 Finish implementing
+  }
+
+  private void preCheckDataReadWrite(
+      JSONObject stargateClaims, List<String> primaryKeyValues, Table tableMetadata)
+      throws JSONException, UnauthorizedException {
+    List<Column> keys = tableMetadata.primaryKeyColumns();
+
+    if (primaryKeyValues.size() > keys.size()) {
+      throw new IllegalArgumentException("Provided more primary key values than exists");
+    }
+
+    for (int i = 0; i < primaryKeyValues.size(); i++) {
+      // If one of the columns exist as a field in the JWT claims and the values do not match then
+      // the request is not allowed.
+      if (stargateClaims.has(STARGATE_PREFIX + keys.get(i).name())) {
+
+        if (!Column.ofTypeText(keys.get(i).type())) {
+          throw new IllegalArgumentException(
+              "Column must be of type text to be used for authorization");
+        }
+
+        String stargateClaimValue = stargateClaims.getString(STARGATE_PREFIX + keys.get(i).name());
+        String columnValue = primaryKeyValues.get(i);
+        if (!stargateClaimValue.equals(columnValue)) {
+          throw new UnauthorizedException("Not allowed to access this resource");
+        }
+      }
+    }
+  }
+
+  private void postCheckDataRead(JSONObject stargateClaims, ResultSet resultSet)
+      throws JSONException, UnauthorizedException {
+    if (resultSet == null || resultSet.rows() == null) {
+      return;
+    }
+
+    // If scopes/fields line up with the resultset then return
+    for (Row row : resultSet.rows()) {
+      for (Column col : row.columns()) {
+        if (stargateClaims.has(STARGATE_PREFIX + col.name())) {
+
+          String stargateClaimValue = stargateClaims.getString(STARGATE_PREFIX + col.name());
+          String columnValue = row.getString(col.name());
+          if (!stargateClaimValue.equals(columnValue)) {
+            throw new UnauthorizedException("Not allowed to access this resource");
```

Comment:
1. Is ordering of iteration important here?
2. If we already fetched the full result set, maybe it would be better to construct an exception with all problematic values and give a more detailed message (instead of `this`). But I am not sure if this is a good solution from a security perspective.

Reply:
> Is ordering of iteration important here?

Actually I could probably flip this around since there will be fewer claims than rows

> If we already fetched the full result set, maybe it would be better to construct an exception with all problematic values and give a more detailed message (instead of this). But I am not sure if this is a good solution from a security perspective.

Correct, from a security standpoint we do not want to leak too much information. Although I'm still trying to decide if it should be an exception or just filter out the things that don't match.